switch from json to ordered_json

### Discussed in https://github.com/nlohmann/json/discussions/3342

<div type='discussions-op-text'>

<sup>Originally posted by **gkriger** February 18, 2022</sup>
From the perspective of casual user the most natural way of switching the code base to ordered_json seems to be replacing
```cpp
using json = nlohmann::json;
```
with
```cpp
using json = nlohmann::ordered_json; 
```
And the things mostly work. However, this:
```cpp
json::object_t jobj({ { "product", "one" } });
```
does not compile however it compiles just fine with using json = nlohmann::json;
Compiler: Microsoft Visual C++ Compiler 16.4.29609.76 (x86_amd64) with c++11 flag.
lib release version 3.10.5
</div>

I can reproduce the issue with Apple Clang:

```
In file included from /Users/niels/Code/projects/json-test/cmake-build-debug/_deps/json-src/include/nlohmann/json.hpp:86:
/Users/niels/Code/projects/json-test/cmake-build-debug/_deps/json-src/include/nlohmann/ordered_map.hpp:38:11: error: no matching constructor for initialization of 'std::vector<std::pair<const string, basic_json<nlohmann::ordered_map>>>'
        : Container{init, alloc} {}
          ^        ~~~~~~~~~~~~~
/Users/niels/Code/projects/json-test/main.cpp:8:20: note: in instantiation of member function 'nlohmann::ordered_map<std::string, nlohmann::basic_json<nlohmann::ordered_map>>::ordered_map' requested here
    json::object_t jobj({ { "product", "one" } });
                   ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:560:5: note: candidate constructor not viable: no known conversion from 'std::initializer_list<basic_json<nlohmann::ordered_map>>' to 'const std::vector<std::pair<const std::string, nlohmann::basic_json<nlohmann::ordered_map>>>' for 1st argument
    vector(const vector& __x, const allocator_type& __a);
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:569:5: note: candidate constructor not viable: no known conversion from 'initializer_list<nlohmann::basic_json<nlohmann::ordered_map>>' to 'initializer_list<std::vector<std::pair<const std::string, nlohmann::basic_json<nlohmann::ordered_map>>>::value_type>' for 1st argument
    vector(initializer_list<value_type> __il, const allocator_type& __a);
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:580:5: note: candidate constructor not viable: no known conversion from 'std::initializer_list<basic_json<nlohmann::ordered_map>>' to 'std::vector<std::pair<const std::string, nlohmann::basic_json<nlohmann::ordered_map>>>' for 1st argument
    vector(vector&& __x, const allocator_type& __a);
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:519:5: note: candidate constructor not viable: no known conversion from 'std::initializer_list<basic_json<nlohmann::ordered_map>>' to 'std::vector<std::pair<const std::string, nlohmann::basic_json<nlohmann::ordered_map>>>::size_type' (aka 'unsigned long') for 1st argument
    vector(size_type __n, const value_type& __x);
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:522:9: note: candidate template ignored: substitution failure [with _InputIterator = std::initializer_list<nlohmann::basic_json<nlohmann::ordered_map>>]: no type named 'reference' in 'std::iterator_traits<std::initializer_list<nlohmann::basic_json<nlohmann::ordered_map>>>'
        vector(_InputIterator __first,
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:537:9: note: candidate template ignored: substitution failure [with _ForwardIterator = std::initializer_list<nlohmann::basic_json<nlohmann::ordered_map>>]: no type named 'reference' in 'std::iterator_traits<std::initializer_list<nlohmann::basic_json<nlohmann::ordered_map>>>'
        vector(_ForwardIterator __first,
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:503:40: note: candidate constructor not viable: requires single argument '__a', but 2 arguments were provided
    _LIBCPP_INLINE_VISIBILITY explicit vector(const allocator_type& __a)
                                       ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:515:14: note: candidate constructor not viable: requires single argument '__n', but 2 arguments were provided
    explicit vector(size_type __n);
             ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:559:5: note: candidate constructor not viable: requires single argument '__x', but 2 arguments were provided
    vector(const vector& __x);
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:566:5: note: candidate constructor not viable: requires single argument '__il', but 2 arguments were provided
    vector(initializer_list<value_type> __il);
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:572:5: note: candidate constructor not viable: requires single argument '__x', but 2 arguments were provided
    vector(vector&& __x)
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:520:5: note: candidate constructor not viable: requires 3 arguments, but 2 were provided
    vector(size_type __n, const value_type& __x, const allocator_type& __a);
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:497:5: note: candidate constructor not viable: requires 0 arguments, but 2 were provided
    vector() _NOEXCEPT_(is_nothrow_default_constructible<allocator_type>::value)
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:530:9: note: candidate constructor template not viable: requires at least 3 arguments, but 2 were provided
        vector(_InputIterator __first, _InputIterator __last, const allocator_type& __a,
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/vector:544:9: note: candidate constructor template not viable: requires at least 3 arguments, but 2 were provided
        vector(_ForwardIterator __first, _ForwardIterator __last, const allocator_type& __a,
        ^
1 error generated.
```

The issue seems to be a missing constructor in the implementation of `ordered_map`.

--- a/include/nlohmann/ordered_map.hpp
+++ b/include/nlohmann/ordered_map.hpp
@@ -34,7 +34,7 @@ template <class Key, class T, class IgnoredLess = std::less<Key>,
     template <class It>
     ordered_map(It first, It last, const Allocator& alloc = Allocator())
         : Container{first, last, alloc} {}
-    ordered_map(std::initializer_list<T> init, const Allocator& alloc = Allocator() )
+    ordered_map(std::initializer_list<value_type> init, const Allocator& alloc = Allocator() )
         : Container{init, alloc} {}
 
     std::pair<iterator, bool> emplace(const key_type& key, T&& t)

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -17070,7 +17070,7 @@ template <class Key, class T, class IgnoredLess = std::less<Key>,
     template <class It>
     ordered_map(It first, It last, const Allocator& alloc = Allocator())
         : Container{first, last, alloc} {}
-    ordered_map(std::initializer_list<T> init, const Allocator& alloc = Allocator() )
+    ordered_map(std::initializer_list<value_type> init, const Allocator& alloc = Allocator() )
         : Container{init, alloc} {}
 
     std::pair<iterator, bool> emplace(const key_type& key, T&& t)

--- a/test/src/unit-regression2.cpp
+++ b/test/src/unit-regression2.cpp
@@ -835,6 +835,18 @@ TEST_CASE("regression tests 2")
 
         CHECK(j.dump() == "[1,4]");
     }
+
+    SECTION("issue #3343 - json and ordered_json are not interchangable")
+    {
+        json::object_t jobj({ { "product", "one" } });
+        ordered_json::object_t ojobj({{"product", "one"}});
+
+        auto jit = jobj.begin();
+        auto ojit = ojobj.begin();
+
+        CHECK(jit->first == ojit->first);
+        CHECK(jit->second.get<std::string>() == ojit->second.get<std::string>());
+    }
 }
 
 DOCTEST_CLANG_SUPPRESS_WARNING_POP

